### PR TITLE
Fix template build to use correct slashes on all platforms

### LINKED ISSUE #4871: Building templates on a non-Windows machine is broken
### Description

When building on a macOS machine, we are using MSBuild to template some .json files. These .json files have backslashes in them `\"`:

https://github.com/dotnet/maui/blob/7b3c187a3955a5afe9c4451f2bfdb1bf70fcd1a8/src/Templates/src/templates/maui-blazor/.template.config/template.json#L24

However, when we run this file through `WriteLinesToFile` task in MSBuild, the backslashes are getting flipped to front slashes:

https://github.com/dotnet/maui/blob/7b3c187a3955a5afe9c4451f2bfdb1bf70fcd1a8/src/Templates/src/Microsoft.Maui.Templates.csproj#L29-L48

In the binlog I see:

image

but that is getting flipped in the parameter to `Lines`:

image

To fix this, we might need to make our own "replacement" task that doesn't flip the slashes.

See https://github.com/dotnet/msbuild/issues/1622 for more details on the underlying MSBuild bug.

### Steps to Reproduce

1. Build `dotnet/maui` on a macOS machine: `dotnet cake --configuration="Release"`
2. Set your $PATH to include `$REPO_ROOT/bin/dotnet`
3. `dotnet new maui`

### Results

```
Error: Failed to load template from /Users/eerhardt/git/maui/bin/dotnet/template-packs/microsoft.maui.templates.6.0.200-dev.nupkg(/content/templates/maui-blazor/.template.config/template.json).
Details: Newtonsoft.Json.JsonReaderException: After parsing a value an unexpected character was encountered: d. Path 'primaryOutputs[0].condition', line 17, position 43.
   at Newtonsoft.Json.JsonTextReader.ParsePostValue(Boolean ignoreComments)
   at Newtonsoft.Json.JsonTextReader.Read()
   at Newtonsoft.Json.Linq.JContainer.ReadContentFrom(JsonReader r, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JContainer.ReadTokenFrom(JsonReader reader, JsonLoadSettings options)
   at Newtonsoft.Json.Linq.JObject.Load(JsonReader reader, JsonLoadSettings settings)
   at Newtonsoft.Json.Linq.JObject.Load(JsonReader reader)
   at Microsoft.TemplateEngine.JExtensions.ReadJObjectFromIFile(IFile file)
   at Microsoft.TemplateEngine.Orchestrator.RunnableProjects.RunnableProjectGenerator.GetTemplatesAndLangpacksFromDir(IMountPoint source, IList`1& localizations)
```

### Version with bug

Unknown/Other (please specify)

### Last version that worked well

Unknown/Other

### Affected platforms

macOS

### Affected platform versions

All macOS versions

### Did you find any workaround?

Use the template packages from a Windows build. Or use a Windows build to make the templated app and copy it to the macOS.

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by Eilon**
Instead of using ItemGroups, which have platform-specific behavior for slashes, it uses a CodeTask to read/write/modify file contents.

### Description of Change

Change template build to work on Mac (it already works on Windows). The problem was that it was treating lines in a text file as MSBuild Items, which can cause unexpected conversions to happen with `\` or `/`. Instead, an inline code task is used to read, modify, and write files, as plain text.

### Issues Fixed

Fixes #4871

cc @eerhardt 

To anyone wishing to review: I'm really not so great at MSBuild. There might be ways to do this MSBuild stuff more efficiently.

### CODE REVIEW

File: src/Templates/src/Microsoft.Maui.Templates.csproj [hunk 1]
```diff
@@ -1,5 +1,28 @@
...
+            File.WriteAllText(
+                OutputFilename,
+                Regex.Replace(File.ReadAllText(InputFilename), MatchExpression, ReplacementText)
+                );
+            Console.WriteLine("Saved modified '{0}' to {1}'.", InputFilename, OutputFilename);
+          ]]>
+      </Code>
+    </Task>
+  </UsingTask>
+
```
Comment (javiercn) [inline src/Templates/src/Microsoft.Maui.Templates.csproj:25]: Does this work on .NET Core MSBuild?

File: src/Templates/src/templates/maui-blazor/.template.config/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+# This file is generated by the build from template.in.json
+# to substitute a product version number.
+template.json
```
Comment (eerhardt) [inline src/Templates/src/templates/maui-blazor/.template.config/.gitignore:3]: Can these `template.json` files be generated in the `obj` directory instead? Then we wouldn't need to do a `.gitignore` or have to exclude `.gitignore` in the Microsoft.Maui.Templates.csproj.

File: src/Templates/src/Microsoft.Maui.Templates.csproj (hunk 1 above)
Comment (Eilon) [reply src/Templates/src/Microsoft.Maui.Templates.csproj]: I believe so. It seemed to work on my Mac, which I think uses that?

File: src/Templates/src/templates/maui-blazor/.template.config/.gitignore (hunk 2 above)
Comment (Eilon) [reply src/Templates/src/templates/maui-blazor/.template.config/.gitignore]: Yep! I just pushed an update for that. I was being lazy 😁

File: src/Templates/src/Microsoft.Maui.Templates.csproj (hunk 1 above)
Comment (eerhardt) [reply src/Templates/src/Microsoft.Maui.Templates.csproj]: `RoslynCodeTaskFactory`'s work on .NET Core. `CodeTaskFactory` don't.

https://docs.microsoft.com/en-us/visualstudio/msbuild/msbuild-roslyncodetaskfactory?view=vs-2022

**Comment by Eilon** [comment]
Should be ready to review. I tested on Windows and Mac and it _seems_ to do everything right. I was able to build and install the templates, and then also create new projects with the templates and they seemed to have the right contents. I also inspected the built NUPKG to see that it had the right content and it seemed fine as well.

**Comment by Eilon** [comment]
Looks like I need to update the mauilib template too.

**Comment by Eilon** [comment]
Tagging @Redth and @mattleibow for review.

File: src/Templates/src/templates/maui-lib/MauiLib1.csproj [hunk 3]
```diff
@@ -1,16 +1,16 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
 	<PropertyGroup>
-		<TargetFrameworks>DOTNET_TFM_VALUE;DOTNET_TFM_VALUE-android;DOTNET_TFM_VALUE-ios;DOTNET_TFM_VALUE-maccatalyst</TargetFrameworks>
-		<TargetFrameworks Condition="$([MSBuild]::IsOSPlatform('windows')) and '$(MSBuildRuntimeType)' == 'Full'">$(TargetFrameworks);DOTNET_TFM_VALUE-windows10.0.19041</TargetFrameworks>
+		<TargetFrameworks>DOTNET_TFM;DOTNET_TFM-android;DOTNET_TFM-ios;DOTNET_TFM-maccatalyst</TargetFrameworks>
```
Comment (eerhardt) [inline src/Templates/src/templates/maui-lib/MauiLib1.csproj:4]: Why this change? I didn't see a reasoning in the description, and I'm not sure how it is connected to the slash issue. Is it just making the parameter naming consistent?

**Review by eerhardt — APPROVED** [review]
LGTM. Thanks for fixing this! Now I can use my build on mac to `dotnet new maui` again. :)

File: src/Templates/src/templates/maui-lib/MauiLib1.csproj (hunk 3 above)
Comment (Eilon) [reply src/Templates/src/templates/maui-lib/MauiLib1.csproj:4]: Ah, this was unifying all the string substitutions to be in one file. It was already done before by editing template.json _and_ editing the csproj. Now it all goes in template.json and gets substituted using the .NET CLI template engine.

File: src/Templates/src/Microsoft.Maui.Templates.csproj [hunk 4]
```diff
@@ -1,5 +1,28 @@
 <Project Sdk="Microsoft.NET.Sdk">
 
+  <UsingTask TaskName="ReplaceFileText" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
+    <ParameterGroup>
+      <InputFilename ParameterType="System.String" Required="true" />
+      <OutputFilename ParameterType="System.String" Required="true" />
+      <MatchExpression ParameterType="System.String" Required="true" />
```
Comment (danmoseley) [inline src/Templates/src/Microsoft.Maui.Templates.csproj:7]: what about making this and the replacement parameters into `string[]` so that you don't need to call it repeatedly below, reading and writing the file each time?

Comment (Eilon) [reply src/Templates/src/Microsoft.Maui.Templates.csproj:7]: If an MSBuild experts wants to offer a way to do that, I'm more than happy. I based this off of copying what another build did because I am a copy & paste expert 😁 I agree it's inefficient to re-save multiple times, but in this part of the build it should not add significant time (some low number of milliseconds, presumably).

Comment (danmoseley) [reply src/Templates/src/Microsoft.Maui.Templates.csproj:7]: I'm an MSBuild expert 👅 

```xml
<Project>


  <UsingTask TaskName="ReplaceFileText" TaskFactory="RoslynCodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.Core.dll">
    <ParameterGroup>
      <InputFilenames ParameterType="System.String[]" Required="true" />
      <OutputFilenames ParameterType="System.String[]" Required="true" />
      <MatchExpression ParameterType="System.String" Required="true" />
      <ReplacementText ParameterType="System.String" Required="true" />
    </ParameterGroup>
    <Task>
      <Using Namespace="System" />
      <Using Namespace="System.IO" />
      <Using Namespace="System.Text.RegularExpressions" />
      <Code Type="Fragment" Language="cs">
        <![CDATA[
            if (InputFilenames.Length != OutputFilenames.Length) throw new Exception("..");

            for (int i = 0; i < InputFilenames.Length; i++)
            {
              File.WriteAllText(
                  OutputFilenames[i],
                  Regex.Replace(File.ReadAllText(InputFilenames[i]), MatchExpression, ReplacementText)
                  );
            }
          ]]>
      </Code>
    </Task>
  </UsingTask>

  <Target Name="t">

    <ReplaceFileText
        InputFilenames="a.txt;b.txt"
        OutputFilenames="a2.txt;b2.txt"
        MatchExpression="WIN2D_VERSION_VALUE"
        ReplacementText="BLAH" />

  </Target>
</Project>
```

this works. 😄

Comment (danmoseley) [reply src/Templates/src/Microsoft.Maui.Templates.csproj:7]: Oh, ha, thats multiple input/output files, you wanted multiple replacements but the principal is exactly the same.

Comment (Eilon) [reply src/Templates/src/Microsoft.Maui.Templates.csproj:7]: I will try to adapt your C# in XML to modify JSON that updates other XML so that people can use C# to suit my needs 😁

Comment (danmoseley) [reply src/Templates/src/Microsoft.Maui.Templates.csproj:7]: ha. Yeah, I'm not sure that the "C# in XML" idea was a great one originally, and I'm guessing that we still don't cache the assemblies either..

Comment (Eilon) [reply src/Templates/src/Microsoft.Maui.Templates.csproj]: I adapted it and it seems to work. Thank you for the suggestion!

**Comment by akoeplinger** [comment]
Another approach than the CodeTask would be using `Microsoft.DotNet.Build.Tasks.Templating` from arcade, here's how we're using it in dotnet/runtime: https://github.com/dotnet/runtime/blob/2330b4bef645bf8bcd3011f35968e55192c697f5/src/installer/pkg/sfx/installers/dotnet-host.proj#L121-L125

**Review by hartez — APPROVED** [review]

**Review by mattleibow — COMMENTED** [review]
This looks very nice. You might be nearing that msbuild-foo level soon.

**Comment by mattleibow** [comment]
> Another approach than the CodeTask would be using `Microsoft.DotNet.Build.Tasks.Templating` from arcade, here's how we're using it in dotnet/runtime: https://github.com/dotnet/runtime/blob/2330b4bef645bf8bcd3011f35968e55192c697f5/src/installer/pkg/sfx/installers/dotnet-host.proj#L121-L125

This looks very interesting...

**Comment by Eilon** [comment]
Thank you everyone for all the feedback. I logged https://github.com/dotnet/maui/issues/5101 to track switching from the CodeTask to the Arcade Templating task. I didn't want to block this PR because it's already functional. Who knows what random issue I'll run into by trying to use the Arcade task 😁
